serializer.h

```diff
@@ -1757,6 +1765,122 @@ void serialize(Archive & archive, const polymorphic_wrapper<Type> & object)
     registry_instance.serialize(archive, *object);
 }
 
+/**
+ * Support types introduced in c++17
+ */
+#if __cplusplus > 201402L
+/**
+ * The base function of the variant serialization helper function.
+ */
+template <typename Archive, typename Variant,
+    typename = typename Archive::loading
+>
+void serialize(Archive &, const Variant &, std::size_t)
+{
+}
+
+/**
+ * Serialize std::variant, operates on loading(input) archives.
+ * Enumerate on all the variant types, until the active type is
+ * reached, and serialize it.
+ */
+template <typename Archive, typename VariantItem, typename... VariantItems,
+    typename Variant, typename = typename Archive::loading
+>
+void serialize(Archive & archive, Variant & variant, std::size_t index)
+{
+    if (0 == index) {
+        VariantItem item;
+        archive(item);
+        variant = item;
+        return;
+    }
+
+    serialize<Archive, VariantItems...>(archive, variant, index - 1);
+}
+
+/**
+ * Serialize std::variant, operates on loading (input) archives.
+*/
+template <typename Archive, typename... VariantItems,
+    typename = typename Archive::loading
+>
+void serialize(Archive & archive, std::variant<VariantItems...> & variant)
+{
+    // Serialize the active variant index
+    decltype(variant.index()) active_index;
+    archive(active_index);
+    if (active_index >= sizeof...(VariantItems)) {
+        throw out_of_range("Variant type mismatch");
+    }
+
+    // Serialize the active variant type, if one exists.
+    variant = {};
+    if (std::variant_npos != active_index) {
+        // A helper function for the variant to trigger the archive()
+        // on the active variant index.
+
+        serialize<Archive, VariantItems...>(archive, variant, active_index);
+    }
+}
+
+/**
+ * Serialize std::variant, operates on saving (output) archives.
+ */
+template <typename Archive, typename... VariantItems,
+    typename = typename Archive::saving
+>
+void serialize(Archive & archive, const std::variant<VariantItems...> & variant)
+{
+    auto variant_index = variant.index();
+
+    // Serialize the variant index, to tell which variant type was active.
+    archive(variant_index);
+
+    // Serialize the active variant type, if one exists.
+    if (std::variant_npos != variant_index) {
+        std::visit([&archive] (auto && value) {
+            archive(value);
+        }, variant);
+    }
+}
+
+/**
+ * Serialize std::optional, operates on loading (input) archives.
+ */
+template <typename Archive, typename OptionalType,
+    typename = typename Archive::loading
+>
+void serialize(Archive & archive, std::optional<OptionalType> & optional)
+{
+    optional = std::nullopt;
+
+    bool has_value;
+    archive(has_value);
+    if (has_value) {
+        OptionalType value;

```

Comment:
I think we need to construct this in a way that a private default constructor works, look at how this is done with "access" structure.

Reply:
Done, not sure if this is what you meant.